### ISSUE TITLE
[BUG] Investigate group-by scan `COUNT` aggregation returning 0-based results

### ISSUE THREAD

**Issue by mythrocks** [rapidsai/cudf#10237]
It appears that the group-by scan aggregation result for `COUNT()` returns a 0-based value. i.e. when the group has 1 row, the `COUNT` seems to return 0.

Consider the JNI [`TableTest::testGroupByScan()`](https://github.com/rapidsai/cudf/blob/branch-22.04/java/src/test/java/ai/rapids/cudf/TableTest.java#L3953):
```java
try (Table t1 = new Table.TestBuilder()
    .column( "1",  "1",  "1",  "1",  "1",  "1",  "1",  "2",  "2",  "2",  "2")
    .column(   0,    1,    3,    3,    5,    5,    5,    5,    5,    5,    5)
// ...
// with `COUNT(*)` aggregation.
    GroupByScanAggregation.count(NullPolicy.INCLUDE).onColumn(2),
// ...
// yields the following results:
   .column(   0,    0,    0,    1,    0,    1,    2,    0,    1,    2,    3) // odd why is this not 1 based?
```
One would have expected the following results instead:
```java
   .column(   1,    1,    1,    2,    1,    2,    3,    1,    2,    3,    4)
```
This might well be by design, but it warrants investigation.

**Comment by jrhemstad** [comment]
It definitely isn't by design.

**Comment by GregoryKimball** [comment]
@mythrocks Would you please make a C++ repro for this bug? Or is it only happening in the JNI layer?

**Comment by davidwendt** [comment]
It does indeed look intentional if I'm reading these gtests right.
https://github.com/rapidsai/cudf/blob/fb703b1a5384606e7fc13e5e931c6f57eb0bb8ff/cpp/tests/groupby/count_scan_tests.cpp#L49-L53
https://github.com/rapidsai/cudf/blob/fb703b1a5384606e7fc13e5e931c6f57eb0bb8ff/cpp/tests/groupby/count_scan_tests.cpp#L146-L151
https://github.com/rapidsai/cudf/blob/fb703b1a5384606e7fc13e5e931c6f57eb0bb8ff/cpp/tests/groupby/count_scan_tests.cpp#L177-L181

**Comment by mythrocks** [comment]
>  It looks like it is waiting for you to have a non-java repro case. Do you need someone else to help you write it?

I have been otherwise occupied, so I didn't get the time to write a repro. I'll get on writing a suitable test.

**Comment by jrhemstad** [comment]
I missed in the original issue description that this was for groupby *scan*. I had thought this was just for regular groupby `count`.

I don't know much about the groupby scan implementation, so it sounds like a bug of using exclusive scan vs inclusive scan.

**Comment by davidwendt** [comment]
This looks like it matches the Pandas groupby cumcount aggregation
https://pandas.pydata.org/docs/reference/api/pandas.core.groupby.DataFrameGroupBy.cumcount.html

Changing this will likely break many upstream functions.
Suggest we close this as won't fix.

**Comment by vyasr** [comment]
Matching pandas is probably what this was designed to do, but FWIW this is not how Polars works either:
```python
In [22]: import polars as pl

In [23]: df = pl.LazyFrame({
    ...:     "a": ["1", "1", "1", "1", "2", "2", "2", "2"],
    ...:     "b": [0, 1, 3, 3, 5, 5, 5, 5],
    ...: })

In [24]: df.with_columns(count=pl.col("b").cum_count().over("b")).collect()
Out[24]:
shape: (8, 3)
┌─────┬─────┬───────┐
│ a   ┆ b   ┆ count │
│ --- ┆ --- ┆ ---   │
│ str ┆ i64 ┆ u32   │
╞═════╪═════╪═══════╡
│ 1   ┆ 0   ┆ 1     │
│ 1   ┆ 1   ┆ 1     │
│ 1   ┆ 3   ┆ 1     │
│ 1   ┆ 3   ┆ 2     │
│ 2   ┆ 5   ┆ 1     │
│ 2   ┆ 5   ┆ 2     │
│ 2   ┆ 5   ┆ 3     │
│ 2   ┆ 5   ┆ 4     │
└─────┴─────┴───────┘
```
This operation isn't currently supported with cudf-polars though.

**Comment by davidwendt** [comment]
How do we resolve this then? 
I'm inclined to change this in libcudf since 1-based makes more sense to me.
The change looks straightforward in libcudf as far as can tell. 
I'm not sure how difficult it would be for the upstream dependencies.

**Comment by vyasr** [comment]
I agree that we should change this in libcudf. 0-based is strange. We can modify cudf's pandas API to insert a binary op subtracting one after the fact.

**Comment by davidwendt** [comment]
Ok, I made the changes to libcudf in #20168 but of course all the python and java tests are failing.
What is the recommended process for this?

**Comment by vyasr** [comment]
Are you asking if someone can help update the Python and Java code to get the tests passing? That shouldn't be a problem if so.

**Comment by davidwendt** [comment]
> Are you asking if someone can help update the Python and Java code to get the tests passing? That shouldn't be a problem if so.

Yes, I definitely need help. 
But what I'm asking about is the process here since it seems this would need to be one PR to make all the changes at once?
Should we just have multiple people pushing to one PR?

**Comment by vyasr** [comment]
Yes, I think so. I think we have to merge all of the fixes atomically to avoid any breakage. I can take a look at the Python failures in your PR.
